src/list.rs

```diff
@@ -248,21 +249,140 @@ impl Drop for ListGuard<'_> {
 }
 
 /// An entry representing a registered listener.
-pub(crate) struct Entry {
-    /// The state of this listener.
-    state: Cell<State>,
+enum Entry {
+    /// Contains the listener state.
+    Listener {
+        /// The state of the listener.
+        state: Cell<State>,
 
-    /// Previous entry in the linked list.
-    prev: Cell<Option<NonZeroUsize>>,
+        /// The previous listener in the list.
+        prev: Cell<Option<NonZeroUsize>>,
 
-    /// Next entry in the linked list.
-    next: Cell<Option<NonZeroUsize>>,
+        /// The next listener in the list.
+        next: Cell<Option<NonZeroUsize>>,
+    },
+
+    /// An empty slot that contains the index of the next empty slot.
+    Empty(NonZeroUsize),
+
+    /// Sentinel value.
+    Sentinel,
+}
+
+impl fmt::Debug for Entry {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        struct TakenState<'a> {
+            state: Option<State>,
+            cell: &'a Cell<State>,
+        }
+
+        impl Drop for TakenState<'_> {
+            fn drop(&mut self) {
+                self.cell.set(self.state.take().unwrap());
```

Comment:
hmm I think it would be nicer to reuse a kind of `State` here as a sentinel/placeholder, e.g. `core::mem::replace(&mut self.state, State::NotifiedTaken)` or such.b